src/FwdState.cc

```diff
@@ -803,17 +804,21 @@ FwdState::noteConnection(HappyConnOpener::Answer &answer)
         syncHierNote(answer.conn, request->url.host());
         Must(!Comm::IsConnOpen(answer.conn));
         answer.error.clear(); // preserve error for errorSendComplete()
+        destinationReceipt = nullptr;
```

Comment:
I think we should remove this line. We should only have code that _sets_ destinationReceipt. Otherwise, we would have to (remember to) add more and more lines like this one, and the code will become inconsistent. It can be argued that it is already inconsistent because we already do not nullify destinationReceipt when the connection is closing() but not `reused`.

How about _asserting_ that destinationReceipt is nil upon entering the noteConnection() method instead? If there are other pending bugs that make such assertion a bad idea, perhaps we can nullify destinationReceipt there with a TODO to assert once those other problems are fixed?

Reply:
> How about asserting that destinationReceipt is nil 

Done.

>  If there are other pending bugs 

I am not aware of such bugs. FwdState::noteConnection() is passed to HappyConnOpener as a callback in FwdState::connectStart() and destinationReceipt is nullified there. HappyConnOpener can call the callback only once (HappyConnOpener::sendFailure() or HappyConnOpener::sendSuccess()) because it resets the callback_ to nil after that. So destinationReceipt should be definitely nil at the time of noteConnection() call.

BTW, we can apply similar logic to FwdState::serverConn, asserting that it is nil in successfullyConnectedToPeer(). I did not change this for now.

Reply:
> I am not aware of such bugs.

I thought there was a bug where we suspected that a second HappyConnOpener was launched when another HappyConnOpener was still running. My recollection may be very wrong on this one though. Just checking -- the assertion is correct with or without that bug.

Reply:
You are right - I forgot about this recently detected [problem](https://measurementfactory.atlassian.net/browse/SQUID-478?focusedCommentId=43671&oldIssueView=true&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-43671). However that wrong second HappyConnOpener instance already causes many other assertions, so I would not remove our new assertion. Do you think we should still replace the assertion with the TODO for now?

Reply:
Probably not. There is no _good_ reason to replace it, especially in the now-official PR. I was just hoping to reduce bug misattribution (i.e. folks blaming the new PR when, in fact, the bug was added in another, older PR). I will add a GitHub comment there.